### PR TITLE
[SPARK-13477][SQL] Expose new user-facing Catalog interface

### PR DESCRIPTION
**Pull request by andrewor14**
## What changes were proposed in this pull request?
#12625 exposed a new user-facing conf interface in `SparkSession`. This patch adds a catalog interface.
## How was this patch tested?

See `CatalogSuite`.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+  def listFunctions(dbName: String): Array[Function]
+
+  /**
+   * Returns a list of columns for the given table in the current database.
+   */
+  @throws[AnalysisException]("table does not exist")
+  def listColumns(tableName: String): Array[Column]
+
+  /**
+   * Returns a list of columns for the given table in the specified database.
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala]: add some annotations

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+/**
+ * Catalog interface for Spark. To access this, use `SparkSession.catalog`.
+ */
+abstract class Catalog {
+
+  /**
+   * Returns the current default database in this session.
+   *
+   * @group ddl_ops
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala]: remove all the group thingy here

Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala]: ok

**Comment by rxin** [comment]
Merging in master.

**Pull request closed without merging**